src/google_publisher.rs

```diff
@@ -0,0 +1,211 @@
+use futures::{TryFutureExt, TryStreamExt};
+use std::{borrow::Cow, future::Future, pin::Pin, sync::Arc, task};
+use yup_oauth2::authenticator::Authenticator;
+
+use crate::{PublisherResult, ValidatedMessage};
+
+#[derive(Debug, thiserror::Error)]
+enum GooglePubsubError {
+    #[error("Could not get authentication token")]
+    GetAuthToken(#[source] yup_oauth2::Error),
+    #[error("Could not POST the request with messages")]
+    PostMessages(#[source] hyper::Error),
+    #[error("Could not construct the request URI")]
+    ConstructRequestUri(#[source] http::uri::InvalidUri),
+    #[error("Could not construct the request")]
+    ConstructRequest(#[source] http::Error),
+    #[error("Could not serialize a publish request")]
+    SerializePublishRequest(#[source] serde_json::Error),
+    #[error("Publish request failed with status code {1}")]
+    ResponseStatus(
+        #[source] Option<Box<dyn std::error::Error + Sync + Send>>,
+        http::StatusCode,
+    ),
+    #[error("Could not parse the response body")]
+    ResponseParse(#[source] serde_json::Error),
+    #[error("Could not receive the response body")]
+    ResponseBodyReceive(#[source] hyper::Error),
+}
+
+/// A publisher that uses the Google Cloud Pub/Sub service as a message transport
+///
+/// # Examples
+///
+/// ```no_run
+/// async {
+///     let google_project =
+///         std::env::var("GOOGLE_CLOUD_PROJECT").unwrap();
+///     let google_credentials = std::env::var("GOOGLE_APPLICATION_CREDENTIALS").unwrap();
+///     let secret = yup_oauth2::read_service_account_key(google_credentials)
+///         .await
+///         .expect("$GOOGLE_APPLICATION_CREDENTIALS is not a valid service account key");
+///     let client = hyper::Client::builder().build(hyper_openssl::HttpsConnector::new()?);
+///     let authenticator = yup_oauth2::ServiceAccountAuthenticator::builder(secret)
+///         .hyper_client(client.clone())
+///         .build()
+///         .await
+///         .expect("could not create an authenticator");
+///     Ok::<_, Box<dyn std::error::Error>>(
+///         hedwig::publishers::GooglePubSubPublisher::new(google_project, client, authenticator)
+///     )
+/// };
+/// ```
+#[allow(missing_debug_implementations)]
+pub struct GooglePubSubPublisher<C>(Arc<PublisherInner<C>>);
+
+struct PublisherInner<C> {
+    google_cloud_project: Cow<'static, str>,
+    client: hyper::Client<C>,
+    authenticator: Authenticator<C>,
+}
+
+impl<C> GooglePubSubPublisher<C> {
+    /// Create a new Google Cloud Pub/Sub publisher
+    pub fn new<P>(
+        project: P,
+        client: hyper::Client<C>,
+        authenticator: Authenticator<C>,
```

Comment:
ehh. why not construct the authenticator in the library? Simplifies the usage..

Reply:
> I also made the library assume less about how people might want to use this. A good example of this is GooglePubSubPublisher accepting an arbitrary oauth2 authenticator (so that different strategies could be chosen by the user; previously we supported service accounts only).

From the PR description.

Reply:
I think we should still provide multiple methods - one simple and one more flexible.

Reply:
Will evaluate this as we integrate this library, but your reasoning is sound.